codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -111,6 +182,65 @@ public void updateStreamableBytes(StreamState state) {
     }
 
     @Override
+    public void updateDependencyTree(int childStreamId, int parentStreamId, short weight, boolean exclusive) {
+        if (weight < MIN_WEIGHT || weight > MAX_WEIGHT) {
+            throw new IllegalArgumentException(String.format(
+                    "Invalid weight: %d. Must be between %d and %d (inclusive).", weight, MIN_WEIGHT, MAX_WEIGHT));
+        }
+        if (childStreamId == parentStreamId) {
+            throw new IllegalArgumentException("A stream cannot depend on itself");
+        }
+
+        State state = state(childStreamId);
+        if (state == null) {
+            if (maxStateOnlySize == 0) {
+                return;
+            }
+            state = new State(childStreamId);
+            stateOnlyRemovalQueue.add(state);
+            stateOnlyMap.put(childStreamId, state);
+        }
+
+        State newParent = state(parentStreamId);
+        if (newParent == null) {
+            if (maxStateOnlySize == 0) {
+                return;
+            }
+            newParent = new State(parentStreamId);
+            stateOnlyRemovalQueue.add(newParent);
+            stateOnlyMap.put(parentStreamId, newParent);
+        }
+
+        // if activeCountForTree == 0 then it will not be in its parent's pseudoTimeQueue and thus should not be counted
+        // toward parent.totalQueuedWeights.
+        if (state.activeCountForTree != 0 && state.parent != null) {
+            state.parent.totalQueuedWeights += weight - state.weight;
+        }
+        state.weight = weight;
+
+        if (newParent != state.parent || (exclusive && newParent.children.size() != 1)) {
+            final List<ParentChangedEvent> events;
+            if (newParent.isDescendantOf(state)) {
+                events = new ArrayList<ParentChangedEvent>(2 + (exclusive ? newParent.children.size() : 0));
+                state.parent.takeChild(newParent, false, events);
+            } else {
+                events = new ArrayList<ParentChangedEvent>(1 + (exclusive ? newParent.children.size() : 0));
+            }
+            newParent.takeChild(state, exclusive, events);
+            notifyParentChanged(events);
+        }
+
+        // The location in the dependency tree impacts the priority in the stateOnlyRemovalQueue map. If we created new
+        // State objects we must check if we exceeded the limit after we inset into the dependency tree to ensure the
+        // stateOnlyRemovalQueue has been updated.
+        while (stateOnlyRemovalQueue.size() > maxStateOnlySize) {
```

Comment:
yes this can happen and is expected. See the `+2 because...` comment above the initialization of `stateOnlyRemovalQueue` and also the comment above this line ... are there additional comments that would be helpful?